### PR TITLE
[Issue #1506] Business Analytics Data Storage ADR

### LINKED ISSUE #1506: [ADR]: Business Analytics Data Storage
### Description

We want to determine the data storage solution for our upcoming business intelligence dashboard project.

This data storage solution will be our canonical source of truth for many types of imported data. These data types will range from thousands of records a month (ex. sprint data), up to potentially millions of records of records (ex. Google Analytics) once this project reaches maturity. The chosen solution should be easy to use, cost-effective, and performant for all of our in-scope data types. 

### Approvers

@widal001 @lucasmbrown-usds @coilysiren @jamesbursa 

### Options

S3 (short-term)
Redshift
Postgres
Snowflake

### Decision Criteria

cost of ownership
ease of use
performance
replicable for outside users (ie can a member of the public run reports)

### Definition of Done

- [ ] The approvers for this decision have been identified (ideally before work on the ADR starts)
- [ ] The ADR is created and stored in `documentation/decisions/adr` with status "Accepted"
- [ ] The ADR has been reviewed and approved by the approvers listed above
- [ ] The ADR satisfies requirements that are outlined in the ADR template
- [ ] Any follow-up tickets have been created (if necessary)

### PR DESCRIPTION
**Pull request by coilysiren**
## Summary

Fixes #1506

### CODE REVIEW

File: documentation/decisions/adr/2024-03-19-dashboard-storage.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  - Grant applications from S3
+- USA spending data CSVs
+- Communication platform stats, eg.
+  - Slack stats
+  - Google Group stats
+  - GitHub stats
+- API application metrics
+- API infrastructure metrics from Cloudwatch
+
+We will not be importing all of these types of data immediately. On the 0 - 6 month timeframe, we will only be importing the smaller datasets (thousands of records). By 2 - 5 years we will be importing all of these types of data, and our data size will be quite large (many millions of records). The desired solutions have different cost/performance characteristics in those time ranges, and we will need to evaluate those differences.
```
Comment (widal001) [inline documentation/decisions/adr/2024-03-19-dashboard-storage.md:26]: One small note on this: I imagine by the 6 month mark we _might_ be ingesting analytics and infrastructure metrics. We'd certainly want to start doing that by the 12-month mark.

I think the biggest outstanding question for those sources of data is whether we want to ingest point level data (e.g. individual page views, clicks, and API calls) or if we'd do some level of aggregation _before_ loading it into our data warehouse.

Just to "show our work" around volume of data metrics:

- ~250k users in the past 7 days (based on latest GA metrics for grants.gov)
- ~4 average page views per user per visit (conservative estimate for sizing purposes)
- 52 weeks per year
- ~52 million page view records per year

Comment (widal001) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:26]: This number could grow geometrically though if we also want to track things like impressions and clicks and tie those to sessions and active devices.

File: documentation/decisions/adr/2024-03-19-dashboard-storage.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+### S3
+
+AWS S3 is a file storage system, traditionally used to store data in the form of individual documents. S3 is best for storing data types formatted as individual documents, or data types where all of the relevant data can be stored within a single file. Due to that, S3 becomes non-ideal for anything in the millions of records, particularly when those records are split up across multiple files. S3 is best in a business intelligence context when used with small datasets, such as our GitHub data. Large data sets require a query layer to be integrated into S3, a query layer like AWS Athena or AWS Redshift. For large enough data sizes, performance starts to become the key issue with S3.
+
+[Data is hosted in S3 at $0.023 per GB-month](https://aws.amazon.com/s3/pricing/).
+
+At 0 - 6 months, S3 is a reasonable choice due to our small data sizes. Past that point, performance issues with large data sizes make S3 a non-ideal choice.
+
+At 2 - 5 years, S3's performance issues require the introduction of another query / compute layer like AWS Athena or AWS Redshift.
```
Comment (widal001) [inline documentation/decisions/adr/2024-03-19-dashboard-storage.md:53]: These kinds of timeline-based comparisons are super helpful @coilysiren !!

File: documentation/decisions/adr/2024-03-19-dashboard-storage.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+- performance
+
+All of these options should be evaluated for 0 - 6 months, 6 months - 2 years, 2 years - 5 years
+
+## Options Considered
+
+- S3
+- Redshift
+- Postgres
+- Snowflake
```
Comment (widal001) [inline documentation/decisions/adr/2024-03-19-dashboard-storage.md:41]: Would it be worth adding Aurora to this list of options per @bretthrosenblatt's comments in Slack?

Comment (bretthrosenblatt) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:41]: Aurora=Postgres in this respect

Comment (bretthrosenblatt) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:41]: I'd recommend both, actually

**Comment by bretthrosenblatt** [comment]
@coilysiren 
There's no benefit to using serverless in a prod environment unless you're sure it would be down at least 80% of the time (completely idle), otherwise it stays at the high range and costs more. For Redshift you'd likely have a much larger instance, and pay for spectrum as well.

**Comment by coilysiren** [comment]
> There's no benefit to using serverless in a prod environment unless you're sure it would be down at least 80% of the time

It's a business analytics data warehouse getting filled by an ELT process, its _**definitely**_ going to down 80% of the time.

**Comment by bretthrosenblatt** [comment]
@coilysiren 

It's a business analytics data warehouse getting filled by an ELT process, its _**definitely**_ going to down 80% of the time.

By down meaning nothing would access, queries, keepalive pings, cleanup actions, etc., and that the scale time would be acceptable

**Comment by bretthrosenblatt** [comment]
@coilysiren 
I don't think we are going to be doing enforcement or validation of most of this data. I believe the plan is to use the ELT tool to pull the data as-is

The 'source of truth' comment threw me off...different meaning for me

**Comment by coilysiren** [comment]
@acouch [for the decision criteria in the ticket](https://github.com/HHS/simpler-grants-gov/issues/1506), you wrote

> replicable for outside users (ie can a member of the public run reports)

Has it been documented from a security impact point of view that we want the public to have _**direct access to the analytics database**_? Because that's the implication of adding this as a requirement here. It's a non-traditional requirement, so I wanted to double-check that I'm understanding this correctly.

**Comment by coilysiren** [comment]
@bretthrosenblatt 

> The potential performance advantage would be irrelevant, so I don't see how a separate platform would be of added benefit to what you already have.

Given that Redshift is an optimized for OLAP queries, I don't understand how this could possibly be the case. Everything I'm reading online says that Redshift is built for the data warehouse use case, but it sounds like you're saying that's wrong?

**Comment by bretthrosenblatt** [comment]
> @bretthrosenblatt
> 
> > The potential performance advantage would be irrelevant, so I don't see how a separate platform would be of added benefit to what you already have.
> 
> Given that Redshift is an optimized for OLAP queries, I don't understand how this could possibly be the case. Everything I'm reading online says that Redshift is built for the data warehouse use case, but it sounds like you're saying that's wrong?

It is designed for OLAP and handles analytics much better, but what you're talking about is mainly in the storage layer. Because it's a col store, every bit of data is fully normalized and compresses far better, and the queries are spread across mpp nodes, so they are dramatically faster on large data sets. The only thing you're missing are the large data sets. Given the sizes discussed here you're unlikely to see much of a difference, and may even be slower, due to the storage and transaction advantage in Aurora. 

Seems like we're trying to change the problem to match a desired solution.

**Comment by widal001** [comment]
> @acouch [for the decision criteria in the ticket](https://github.com/HHS/simpler-grants-gov/issues/1506), you wrote
> 
> > replicable for outside users (ie can a member of the public run reports)
> 
> Has it been documented from a security impact point of view that we want the public to have _**direct access to the analytics database**_? Because that's the implication of adding this as a requirement here. It's a non-traditional requirement, so I wanted to double-check that I'm understanding this correctly.

Just wanted to add a quick note here, when we expose our metrics and underlying data it will be through an analytics API, so the public wouldn't have direct access to the OLAP database. We will also most likely will be limiting public access to analytics data to aggregates rather than point-level data for things like site traffic, API calls, etc.

File: documentation/decisions/adr/2024-03-19-dashboard-storage.md (hunk 3 above)
Comment (widal001) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:43]: Gotcha, I wasn't sure if postgres meant RDS -- that's what we're using for our main database right?

Comment (coilysiren) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:43]: @widal001 AWS RDS Aurora PostgreSQL is its full name, and its our main database yes

Comment (bretthrosenblatt) [reply documentation/decisions/adr/2024-03-19-dashboard-storage.md:43]: > Gotcha, I wasn't sure if postgres meant RDS -- that's what we're using for our main database right?

Purposely confusing. There is RDS Postgres, which is cloud-hosted Postgres, and Postgres-compatible Aurora, which is a Postgres-like interface on an Aurora db engine and storage cluster (what you're using)

**Comment by bretthrosenblatt** [comment]
I confirmed the scaling issue for Redshift serverless. You need to set the base RPU (Redshift Processing Unit) to whatever is required for the ETL process, and then if you want the ability to scale up you can set a range. The the scaling is concurrency scaling (read only), and is driven primarily by the number of queries, not the complexity. If you want the opposite you can do so in the Redshift API (increase RPU before ETL, reduce after, or even spin up a temp cluster), or you can just use a price ratio and AWS uses ML to control the workloads. Not sure if this would be ultimately cheaper than a reserved instance though. Looking at a minimum 2.88/hr (just RPU), max likely 46/hr, based on projected work loads (assume 128RPU max, based on average usage reporting)

**Comment by coilysiren** [comment]
> I confirmed the scaling issue for Redshift serverless...

This seems like less of a "scaling issue" and more "the fine grain details about how it scales". I don't see any issues with it.

File: documentation/decisions/adr/2024-03-19-dashboard-storage.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+At 2 - 5 years, performance bottle-necks with S3 mean that Redshift easily wins any comparison.
+
+### Postgres
+
+Postgres, as hosted by AWS RDS, is a SQL database that you can use to store large data sets. Many of the characteristics that apply to Redshift apply to Postgres, with the caveat that Postgres is not built for business intelligence use cases. We have an existing Postgres database, but we would not want to re-use it for our data analytics purposes.
+
+[Data is hosted in Postgres at $0.115 per GB-month](https://aws.amazon.com/rds/postgresql/pricing/), higher than S3 and Redshift.
+
+In every time range, Postgres loses to Redshift due to Postgres being an OLTP database built for real-time data processing. In the 2 - 5 year range, Postgres also loses due to its high GB-month hosting cost. There is however, evidence that Postgres is faster than Redshift, and that Redshift's OLAP advantages don't kick in until the data grows to terabtyes in size.
```
Comment (acouch) [inline documentation/decisions/adr/2024-03-19-dashboard-storage.md:75]: Pros of postgres include:

* Open Source
* Keep stack simplified as we already use it
* Easier to setup locally for open source devs or local dev and for other teams to adopt 
* I'm still skeptical that we will ever need to run queries that require an OLTP database
* We could use the same database for our open source BI tool (Metabase/Superset use postgres)

**Comment by coilysiren** [comment]
Given the evidence and discussion, I've changed the decision in this ADR to instead be in favor of **Postgres**

**Review by acouch — APPROVED** [review]

**Review by bretthrosenblatt — APPROVED** [review]

**Pull request merged**